src/crimson/os/seastore/seastore.h

```diff
@@ -80,375 +91,501 @@ class SeaStore final : public FuturizedStore {
-  tm_ret _setattrs(
-    internal_context_t &ctx,
-    OnodeRef &onode,
-    std::map<std::string,bufferlist>&& aset);
-  tm_ret _rmattr(
-    internal_context_t &ctx,
-    OnodeRef &onode,
-    std::string name);
-  tm_ret _rmattrs(
-    internal_context_t &ctx,
-    OnodeRef &onode);
-  tm_ret _xattr_rmattr(
-    internal_context_t &ctx,
-    OnodeRef &onode,
-    std::string &&name);
-  tm_ret _xattr_clear(
-    internal_context_t &ctx,
-    OnodeRef &onode);
-  tm_ret _create_collection(
-    internal_context_t &ctx,
-    const coll_t& cid, int bits);
-  tm_ret _remove_collection(
-    internal_context_t &ctx,
-    const coll_t& cid);
-  using omap_set_kvs_ret = tm_iertr::future<>;
-  omap_set_kvs_ret _omap_set_kvs(
-    OnodeRef &onode,
-    const omap_root_le_t& omap_root,
-    Transaction& t,
-    omap_root_le_t& mutable_omap_root,
-    std::map<std::string, ceph::bufferlist>&& kvs);
-
-  boost::intrusive_ptr<SeastoreCollection> _get_collection(const coll_t& cid);
-
-  static constexpr auto LAT_MAX = static_cast<std::size_t>(op_type_t::MAX);
-  struct {
-    std::array<seastar::metrics::histogram, LAT_MAX> op_lat;
-  } stats;
-
-  seastar::metrics::histogram& get_latency(
-      op_type_t op_type) {
-    assert(static_cast<std::size_t>(op_type) < stats.op_lat.size());
-    return stats.op_lat[static_cast<std::size_t>(op_type)];
+  seastar::future<> write_meta(
+    const std::string& key,
+    const std::string& value) final {
+    ceph_assert(seastar::this_shard_id() == primary_core);
+    return shard_stores.local().write_meta(
+      key, value).then([this, key, value] {
+      return mdstore->write_meta(key, value);
+    }).handle_error(
+      crimson::ct_error::assert_all{"Invalid error in SeaStore::write_meta"}
+    );
   }
 
-  void add_latency_sample(op_type_t op_type,
-       std::chrono::steady_clock::duration dur) {
-    seastar::metrics::histogram& lat = get_latency(op_type);
-    lat.sample_count++;
-    lat.sample_sum += std::chrono::duration_cast<std::chrono::milliseconds>(dur).count();
+  seastar::future<std::tuple<int, std::string>> read_meta(const std::string& key) final;
+
+  SeaShardStore& get_local_shard_store() final {
+    return shard_stores.local();
   }
-  seastar::metrics::metric_group metrics;
-  void register_metrics();
+
+  static col_obj_ranges_t
+  get_objs_range(CollectionRef ch, unsigned bits);
+
+private:
   seastar::future<> write_fsid(uuid_d new_osd_fsid);
+
+  seastar::future<> prepare_meta(uuid_d new_osd_fsid);
+
   seastar::future<> _mkfs(uuid_d new_osd_fsid);
+
+  seastar::future<> shard_mkfs(
+    std::vector<secondary_device_set_t> &sds,
+    uuid_d new_osd_fsid);
+
+  seastar::future<> sec_shard_mkfs(
+    const std::string path,
+    device_type_t dtype,
+    device_id_t id,
+    std::vector<secondary_device_set_t> &sds,
+    uuid_d new_osd_fsid);
+
+private:
+  std::string root;
+  MDStoreRef mdstore;
+  DeviceRef device;
+  const uint32_t max_object_size = 0;
+  bool is_test;
```

Comment:
Looks root/device/max_object_size/is_test are no longer necessary in `SeaStore` with the above suggestions.